engine/linksystem.go

```diff
@@ -188,7 +188,12 @@ func (e *Engine) generateChunks(mhIter provider.MultihashIterator) (ipld.Link, e
 	// If the cache was resized to expand beyond its original capacity, then
 	// set its size to only as big as the number of links in this list.
 	if resized {
-		ls.Resize(context.Background(), ls.Len())
+		newSize := e.linkCacheSize
+		if chunkCount > e.linkCacheSize {
+			newSize = chunkCount
+		}
```

Comment:
The line
```go
ls.Resize(context.Background(), ls.Len())
```
Was resizing the cache (`ls') down to the number of items current stored, because the rest of the capacity is not needed.

Your change sets the size back to the original size, which will have the effect of immediately removing chunks that were just created and put into the cache.

I don think we ever want to resize back down to the original size.  The cache size should be whatever is needed to hold the last generated set of chunks.

Reply:
> Was resizing the cache (ls) down to the number of items current stored, because the rest of the capacity is not needed.

That line always set the size to the number of items in the cache; which means none of the items in it ever get evicted and it grows indefinitely unless I am missing something. See https://github.com/filecoin-project/index-provider/pull/168/commits/c432f6dea45c704e34e6e601507207bd8881374f for local reproduction of cache growing as large as all the entries ever generated.

> The cache size should be whatever is needed to hold the last generated set of chunks.

Assuming the resizing we are looking for here should work as the comments added above the if statement, then the cache _capacity_ should either be as large as the configured capacity or as large as the longest generated chain which I think the code here does.


Let's discuss this further in a sync call.